### PR TITLE
reenable ubuntu's toolchain ppa for c++20 support

### PR DESCRIPTION
**Pull request by cbodley**
we've used this repo in the past to get access to newer compiler versions. reenable it for ubuntu focal for access to gcc 11

note that these changes don't preserve the original behavior against old distros like trusty/xenial. if there is still interest in building against those releases, i can work through these differences:

trusty: PPA 7 -> 11
xenial: PPA 8 -> 11
bionic: does not use PPA (unchanged)

### CODE REVIEW

**Comment by cbodley** [comment]
i also see some logic around the gcc toolset in https://github.com/ceph/ceph-build/blob/master/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml#L176-L182. shall i raise `gcc_toolset_ver` to 10 there?

File: scripts/build_utils.sh
Comment (tchaikov) [inline scripts/build_utils.sh:655]: just need to make sure that https://tracker.ceph.com/issues/50218 does not apply to focal.

File: scripts/build_utils.sh
Comment (tchaikov) [inline scripts/build_utils.sh:1342]: probably we don't need SCL for accessing GTS-11. see https://centos.pkgs.org/8-stream/centos-appstream-x86_64/gcc-toolset-11-build-11.0-0.el8.x86_64.rpm.html

what we need is but to enable the "AppStream" repo. and
```sh
dnf install gcc-toolset-11-build
```

Comment (cbodley) [reply scripts/build_utils.sh:1342]: ok, thanks. i was being conservative by requiring gcc 10.2, but i can update https://github.com/ceph/ceph/pull/45133 with `BuildRequires: gcc-toolset-11-gcc-c++`

does that mean ubuntu should use_ppa 11 to match?

i'm not familiar with AppStream, and grep didn't find anything under ceph or ceph-build; is that repo expected to be enabled by default, or do i need something extra here?

File: scripts/build_utils.sh
Comment (cbodley) [reply scripts/build_utils.sh:655]: good to know; if that shows up, it will be immediately obvious from teuthology testing

File: scripts/build_utils.sh
Comment (cbodley) [reply scripts/build_utils.sh]: ah, i see that ceph.spec.in has `. /opt/rh/gcc-toolset-9/enable` for this, i'll make that unconditional and raise it to 11

**Review by tchaikov — APPROVED** [review]
lgtm. need to keep an eye on the build status after merging this change though.

**Comment by tchaikov** [comment]
> i also see some logic around the gcc toolset in https://github.com/ceph/ceph-build/blob/master/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml#L176-L182. shall i raise `gcc_toolset_ver` to 10 there?

would be great if you could. probably should point clang to gcc11 since we have gcc-11 here.

BTW, now that we have llvm-toolset-12 in the official repo, probably it's a good opportunity for us to ditch my personal copr repo and start using clang-12 from AppStream! but don't worry! as it's not in the scope of this change, i will take care of this. clang is being used by the ceph-pull-requests jenkins job for a better compile test for the tree.

i will implement this change this week end.

**Comment by cbodley** [comment]
note that octopus doesn't build cleanly with gcc11, as discussed in backport pr https://github.com/ceph/ceph/pull/45206#issuecomment-1057094145. i'm not sure whether this affects pacific yet

we might need to hold off on this until we either resolve those build issues, or use an earlier version for octopus

**Comment by cbodley** [comment]
> i'm not sure whether this affects pacific yet

yes, pacific PRs are also failing in spdk on nodes that enabled gcc11 - ex. https://github.com/ceph/ceph/pull/45245

**Comment by cbodley** [comment]
> yes, pacific PRs are also failing in spdk on nodes that enabled gcc11 - ex. [ceph/ceph#45245](https://github.com/ceph/ceph/pull/45245)

i updated `use_ppa()` so that it hopefully only applies to 'reef' and later, so we shouldn't see these failures against octopus/pacific/quincy

**Comment by tchaikov** [comment]
@cbodley can we drop the "do-not-merge" label? i think it's ready to go.

File: scripts/build_utils.sh
Comment (tchaikov) [inline scripts/build_utils.sh:641]:
```suggestion
        17.2*) # q
```

as reef is not tagged yet. the commits in main are all named something like `v17.2.0-267-g8bae27e49f9`,

```console
$ git describe --tag
v17.2.0-267-g8bae27e49f9
```

and now that quincy is released, we are not likely to build for 17.1.x anymore. so `17.2*` is a safe choice.

**Comment by tchaikov** [comment]
after applying the proposed change, i updated the job of `ceph-dev-new-build` . and the tree built fine: https://shaman.ceph.com/builds/ceph/wip-cxx20-fixes-kefu-1/377ef0a88af30c85a507c9d7fa4d62e5fc14b4ef/

**Comment by cbodley** [comment]
addressed review comment and removed the dnm

**Comment by cbodley** [comment]
the use of `WITH_STATIC_LIBSTDCXX` in ubuntu builds is causing radosgw to crash on startup: https://tracker.ceph.com/issues/57050

**Comment by cbodley** [comment]
looks like we can resolve it with https://github.com/ceph/ceph/pull/47504

**Pull request merged**
